crates/net/network/src/session/conn.rs

```diff
@@ -0,0 +1,156 @@
+//! Connection types for a session
+
+use futures::{Sink, Stream};
+use reth_ecies::stream::ECIESStream;
+use reth_eth_wire::{
+    errors::EthStreamError,
+    message::EthBroadcastMessage,
+    multiplex::{ProtocolProxy, RlpxSatelliteStream},
+    EthMessage, EthStream, EthVersion, P2PStream,
+};
+use reth_net_common::bandwidth_meter::MeteredStream;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::net::TcpStream;
+
+/// The type of the underlying peer network connection.
+pub type EthPeerConnection = EthStream<P2PStream<ECIESStream<MeteredStream<TcpStream>>>>;
+
+/// Various connection types that at least support the ETH protocol.
+pub type EthSatelliteConnection =
+    RlpxSatelliteStream<ECIESStream<MeteredStream<TcpStream>>, EthStream<ProtocolProxy>>;
+
+/// Connection types that support the ETH protocol.
+///
+/// Either a [`EthPeerConnection`] or an [`EthSatelliteConnection`].
+// This type is boxed because the underlying stream is ~6KB,
+// mostly coming from `P2PStream`'s `snap::Encoder` (2072), and `ECIESStream` (3600).
+#[derive(Debug)]
+pub enum EthRlpxConnection {
+    /// A That only supports the ETH protocol.
+    EthOnly(Box<EthPeerConnection>),
+    /// A connection that supports the ETH protocol and __at least one other__ RLPx protocol.
+    Satellite(Box<EthSatelliteConnection>),
+}
```

Comment:
is starting a `SatelliteStream` that much overhead when no extra protocols are shared that this enum is necessary?

Reply:
possibly, because this would involve routing through a channel,